### PR TITLE
Rr/anndemo_topK changes

### PR DESCRIPTION
**Pull request by rrawther**
Changes to support topK inference results

### CODE REVIEW

File: utils/annInferenceApp/inference_receiver.cpp [hunk 1]
```diff
@@ -160,6 +166,48 @@ void inference_receiver::run()
...
+                    std::vector<int> labels;
+                    std::vector<float> probVec;
+                    for(int i = 0; i < count; i++) {
+                        int tag = cmd.data[2 + item_size*i + 0];
+                        imageIndex.push_back(tag);
+                        for (int j=1; j<=top_k; j++){
+                            int label = cmd.data[2 + i*item_size + j];      // label has both label and prob
+                            float prob =  (label>>16)*(1.0f/(float)32767.0f);
+                            prob = std::min(prob, 1.0f);
+                            labels.push_back(label & 0xFFFF);
```
Comment (rgiduthuri) [inline utils/annInferenceApp/inference_receiver.cpp]: Please remove the `prob = std::min(prob, 1.0f);` statement. Its existence can make it hard to catch some bugs.

File: utils/annInferenceServer/inference.cpp [hunk 2]
```diff
@@ -15,12 +15,20 @@
...
+       [&v](size_t i1, size_t i2) {return v[i1] > v[i2];});
+}
+
 InferenceEngine::InferenceEngine(int sock_, Arguments * args_, std::string clientName_, InfComCommand * cmd)
     : sock{ sock_ }, args{ args_ }, clientName{ clientName_ },
       GPUs{ cmd->data[1] },
       dimInput{ cmd->data[2], cmd->data[3], cmd->data[4] },
       dimOutput{ cmd->data[5], cmd->data[6], cmd->data[7] },
-      receiveFileNames { (bool)cmd->data[8] },
+      receiveFileNames { (bool)cmd->data[8] }, topK { cmd->data[9] },
```
Comment (rgiduthuri) [inline utils/annInferenceServer/inference.cpp]: When cmd->data[9] is ZERO, the topK should default to ONE. So, this needs: `topK { cmd->data[9] == 0 ? 1 : cmd->data[9] }`

File: utils/annInferenceServer/inference.cpp [hunk 3]
```diff
@@ -686,31 +701,66 @@ int InferenceEngine::run()
...
+                        ERRCHK(sendCommand(sock, cmd, clientName));
+                        resultCountAvailable -= resultCount;
+                        ERRCHK(recvCommand(sock, cmd, clientName, INFCOM_CMD_INFERENCE_RESULT));
+                    }
+                    if(endOfSequence) {
+                        break;
+                    }
+                }else {
+                    // send topK labels
+                    int maxResults = (INFCOM_MAX_IMAGES_PER_PACKET*2)/(topK+1);
```
Comment (rgiduthuri) [inline utils/annInferenceServer/inference.cpp]: The `INFCOM_MAX_IMAGES_PER_PACKET` was created assuming that top_k=1. This should now be renamed to `INFCOM_MAX_IMAGES_FOR_TOP1_PER_PACKET`.

File: utils/annInferenceServer/inference.cpp [hunk 4]
```diff
@@ -217,9 +224,9 @@ vx_status InferenceEngine::DecodeScaleAndConvertToTensor(vx_size width, vx_size
         img += 12;
     }
     for (; i < length; i++, img += 3) {
-        B_buf[i] = (img[0] * preprocessMpy[0]) + preprocessAdd[0];
-        G_buf[i] = (img[1] * preprocessMpy[1]) + preprocessAdd[1];
-        R_buf[i] = (img[2] * preprocessMpy[2]) + preprocessAdd[2];
+        *B_buf++ = (img[0] * preprocessMpy[0]) + preprocessAdd[0];
+        *G_buf++ = (img[1] * preprocessMpy[1]) + preprocessAdd[1];
+        *R_buf++ = (img[2] * preprocessMpy[2]) + preprocessAdd[2];
```
Comment (rgiduthuri) [inline utils/annInferenceServer/inference.cpp]: To avoid the merge conflicts, please revert this change and merge 'develop' branch into this branch.

File: utils/annInferenceApp/infcom.h [hunk 5]
```diff
@@ -41,13 +41,15 @@
...
 //  * server: InfComCommand:INFCOM_CMD_SEND_IMAGES with data={maxCount}
 //    client: InfComCommand:INFCOM_CMD_SEND_IMAGES with data={count} -- count:0..maxCount
 //    client: for each image: { <tag:32-bit> <size:32-bit> <byte-stream> <eof-marker:32-bit> }
 //            or {<tag:32-bit> <size:32-bit> bytearray(filename)> <eof-marker:32-bit> } if rf == 1
 //  * server: InfComCommand:INFCOM_CMD_INFERENCE_RESULT data={imageCount,0,<tag1>,<label1>,<tag2>,<label2>,...} upto 14 tags
 //    client: InfComCommand:INFCOM_CMD_INFERENCE_RESULT data={imageCount,0,<tag1>,<label1>,<tag2>,<label2>,...} upto 14 tags
+//  * server: InfComCommand:INFCOM_CMD_TOPK_INFERENCE_RESULT data={imageCount,top_k,<tag0>,<label0:prob0>,..<labelk:probk>,<tag1><label0:prob0>..<labelk:probk>, ...} upto imageCount
+//    client: InfComCommand:INFCOM_CMD_TOPK_INFERENCE_RESULT data={imageCount,top_k,<tag0>,<label0:prob0>,..<labelk:probk>,<tag1><label0:prob0>..<labelk:probk>, ...} upto imageCount
```
[the remark this answers is not in the document]
Comment (rrawther) [reply utils/annInferenceApp/infcom.h]: haven't heard of UQ1.15. when searching online, it says means 1 integer and 15 fractional bits ( where as what is used is 15 fractional bits).

File: utils/annInferenceApp/inference_receiver.cpp (hunk 1 above)
Comment (rrawther) [reply utils/annInferenceApp/inference_receiver.cpp]: I am not sure why? I have added this based on our previous discussion

File: utils/annInferenceServer/inference.cpp [hunk 6]
```diff
@@ -1162,16 +1212,35 @@ void InferenceEngine::workDeviceOutputCopy(int gpu)
...
+                sort_indexes(prob_vec, idx);            // sort indeces based on prob
+                std::vector<int>    labels;
+                //labels.push_back(tag);
+                outputQ.enqueue(std::tuple<int,int>(tag,idx[0]));
+                int j=0;
+                for (auto i: idx) {
+                  // make label which is index and prob
+                    //cout<<"tag: "<< tag << "top_" << j << "label and prob: " << i << " " << prob_vec[i] << endl;
+                    int label = (i&0xFFFF)|(((unsigned int)(prob_vec[i]*0x7FFF))<<16);   // convert prob to 16bit float and store in MSBs
+                    labels.push_back(label);
```
[the remark this answers is not in the document]
Comment (rrawther) [reply utils/annInferenceServer/inference.cpp]: the Softmax output is already clamped to values between 0 and 1 which is why I didn't bother to clamp again.

File: utils/annInferenceApp/annInferenceApp.py [hunk 7]
```diff
@@ -297,6 +305,37 @@ def runInference(host,port,GPUs,model,imageDirPath,imageFileList,synsetFileName,
...
+							resultList[tag] = topLabel
+                    if fp:
+                        fp.write(line + '\n')
+                    if verbose or not fp:
+                        print(line)
+                else:
+                    print('RECEIVED INCORRECT RESULT: ', tag, topLabel)
+            if resultCount >= len(imageFileList):
+                sendpkt(sock,(INFCOM_MAGIC,INFCOM_CMD_DONE,(0,),''))
+                break
```
[the remark this answers is not in the document]
Comment (rrawther) [reply utils/annInferenceApp/annInferenceApp.py]: done.

File: utils/annInferenceApp/infcom.h [hunk 8]
```diff
@@ -41,13 +41,15 @@
 // Inference Run Protocol:
 //    client: (connect)
 //  * server: InfComCommand:INFCOM_MODE_INFERENCE
-//    client: InfComCommand:INFCOM_CMD_SEND_MODE with data={INFCOM_MODE_INFERENCE,GPUs,iw,ih,ic,ow,oh,oc,rf} message="modelName [options]", where rf==1 indicates clients is sending filenames
+//    client: InfComCommand:INFCOM_CMD_SEND_MODE with data={INFCOM_MODE_INFERENCE,GPUs,iw,ih,ic,ow,oh,oc,rf, top_k} message="modelName [options]", where rf==1 indicates client sending filenames & top_k for topK inference results
 //  * server: InfComCommand:INFCOM_CMD_SEND_IMAGES with data={maxCount}
```
[the remark this answers is not in the document]
Comment (rrawther) [reply utils/annInferenceApp/infcom.h]: when topK is not specified or ZERO, the behavior is same as before (default).

File: utils/annInferenceServer/inference.cpp (hunk 4 above)
Comment (rrawther) [reply utils/annInferenceServer/inference.cpp]: looks like develop has already these changes. So won't take effect.

File: utils/annInferenceServer/inference.cpp (hunk 2 above)
Comment (rrawther) [reply utils/annInferenceServer/inference.cpp]: I have changed topK ==1 to use the new code where it returns prob and label. So the above change is not needed.

**Comment by rrawther** [comment]
I have committed the changes requested. Please merge the pull request
